src/engine/gnc-pricedb.c

```diff
@@ -1345,68 +1356,147 @@ pricedb_remove_foreach_pricelist (gpointer key,
+
+    recurrance_date_next = recurrance_date_old = day_of_week_date;
+
+    while (g_date_compare (&period_end, &recurrance_date_next) > 0)
+    {
+        gchar *date_str;
+        
+        recurrenceNextInstance (r, &recurrance_date_old, &recurrance_date_next);
+        date_str = g_strdup_printf ("%d/%d/%d", g_date_get_day (&recurrance_date_old),
+                   g_date_get_month (&recurrance_date_old), g_date_get_year (&recurrance_date_old));
+
+        DEBUG("Recurrance date old string for hash table insert is %s", date_str);
+
+        g_hash_table_insert (hash_dates, date_str, "Keep"); // date_str freed by hashtable
+        recurrance_date_old = recurrance_date_next;
+    }
+    g_free (r);
+}
 
 gboolean
-gnc_pricedb_remove_old_prices(GNCPriceDB *db,
-                              Timespec cutoff,
-                              gboolean delete_user,
-                              gboolean delete_last)
+gnc_pricedb_remove_old_prices (GNCPriceDB *db, GList *comm_list,
+                              Timespec first, Timespec cutoff,
+                              PriceRemoveOptions source,
+                              PriceRemoveOptions leave)
 {
+    GHashTable *hash_dates = NULL;
     remove_info data;
     GSList *item;
+    GList  *node;
 
     data.db = db;
     data.cutoff = cutoff;
-    data.delete_user = delete_user;
-    data.delete_last = delete_last;
     data.list = NULL;
+    data.delete_user = FALSE;
+    data.delete_all = FALSE;
+
+    ENTER("Remove Prices for Source %d, leaving %d", source, leave);
 
-    ENTER("db %p, delet_user %d, delete_last %d", db, delete_user, delete_last);
+    // setup the options
+    if (source == PRICE_REMOVE_ALL)
+        data.delete_all = TRUE;
+
+    if (source == PRICE_REMOVE_USER)
+        data.delete_user = TRUE;
+
+    // Walk the list of commodities
+    for (node = g_list_first (comm_list); node; node = g_list_next (node))
     {
-        gchar buf[40];
-        gnc_timespec_to_iso8601_buff(cutoff, buf);
-        DEBUG("checking date %s", buf);
+        GHashTable *currencies_hash = g_hash_table_lookup (db->commodity_hash, node->data);
+        g_hash_table_foreach (currencies_hash, pricedb_remove_foreach_pricelist, &data);
     }
 
-    /* Traverse the database once building up an external list of prices
-     * to be deleted */
-    g_hash_table_foreach(db->commodity_hash,
-                         pricedb_remove_foreach_currencies_hash,
-                         &data);
-
     if (data.list == NULL)
+    {
+        LEAVE("Empty price list");
         return FALSE;
+    }
+    DEBUG("Number of Prices in list is %d", g_slist_length (data.list));
+
+    if (leave != PRICE_REMOVE_DEFAULT)
+    {
+        GDate period_begin = timespec_to_gdate (first);
+        GDate period_end = timespec_to_gdate (cutoff);
+        hash_dates = g_hash_table_new_full (g_str_hash, g_str_equal, free_data, NULL);
```

Comment:
In addition to the comment made earlier about the value of the hash table, I think it would be more efficient to make the key a guint32 and use the result of g_date_get_julian as keys.
That will
* avoid the need for a free function
* allow direct insertions instead of constructing a date string to insert